### PR TITLE
feat/circle-ci-npm-deploy (dry run)

### PR DESCRIPTION
**Pull request by dickie81**
## Description:
Adds dry run semantic release & commitlint for progression towards continuous deployment.

### Checklist for this pull request

* [X] PR is scoped to one task
* [X] Tests are included
* [X] Documentation included
* [X] One of:
  * [X] The PR is not making any UI change
  * [ ] The PR is making a UI change but not a product change
    * [ ] Screenshots included

Thank you!

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -1,35 +1,84 @@
+defaults: &defaults
```
Comment (ranmeirman) [inline circle.yml]: I think that according to their docs it should be under `./circleci/config.yml`

File: circle.yml [hunk 2]
```diff
@@ -1,35 +1,84 @@
...
+
 version: 2
 jobs:
   build:
-    docker:
-      - image: applicaster/zapp:0.8
+    <<: *defaults
     steps:
       - restore_cache:
           key: react-native-zapp-bridge-{{ .Branch }}-{{ .Revision }}
```
Comment (ranmeirman) [inline circle.yml]: I would add version like`-v1`, without it, it will be less intuitive to invalidate cache

File: circle.yml [hunk 3]
```diff
@@ -1,35 +1,84 @@
...
           keys:
             - react-native-zapp-bridge-{{ .Branch }}-{{ .Revision }}
             - react-native-zapp-bridge-{{ .Branch }}-
             - react-native-zapp-bridge-
+
       - checkout
+
       - save_cache:
           key: react-native-zapp-bridge-{{ .Branch }}-{{ .Revision }}
```
Comment (ranmeirman) [inline circle.yml]: same as above

File: circle.yml [hunk 4]
```diff
@@ -1,35 +1,84 @@
...
+  deploy:
+    <<: *defaults
+
+    steps:
+      - attach_workspace:
+          at: ~/repo
+
+      - run:
+          name: Semantic release
+          command: npx semantic-release --dry-run
```
Comment (ranmeirman) [inline circle.yml]: for the dry run its ok, but you will probably want to publish only when pushing a specific branch

**Review by ranmeirman — CHANGES_REQUESTED** [review]
wrote some comments

File: circle.yml (hunk 1 above)
Comment (francois-codes) [reply circle.yml]: `./.circleci/config.yml`

File: circle.yml (hunk 4 above)
Comment (francois-codes) [reply circle.yml]: I agree, but we should define a way to publish as release candidate to - maybe not each branch, but it could be useful

**Review by francois-codes — CHANGES_REQUESTED** [review]
Nothing special on top of @ranmeirman 's comments - except that it would be great to add the commit structure to the readme

File: circle.yml (hunk 4 above)
Comment (dickie81) [reply circle.yml]: branches are filtered below so only changes to master will result in publish jobs

File: circle.yml (hunk 2 above)
Comment (dickie81) [reply circle.yml]: I'm not sure I understand;
`react-native-zapp-bridge-{{ .Branch }}-{{ .Revision }}-v1`?
How does it make it more intuitive?

**Comment by dickie81** [comment]
@f-roland Can we think about auto RCs on a new PR? Maybe merges into `release-candidate` branch?

File: circle.yml (hunk 1 above)
Comment (dickie81) [reply circle.yml]: done

File: circle.yml (hunk 2 above)
Comment (francois-codes) [reply circle.yml]: In ci 2.0 the cache mechanism weirdly ignores ‘rebuild without cache’
So if you change this file to check a cache with `v2` key, it won’t find it, so we have a reliable way to invalidate cache on a branch / commit

File: circle.yml (hunk 4 above)
Comment (francois-codes) [reply circle.yml]: Sounds good, please add it to readme as well :)

**Comment by dickie81** [comment]
Also @f-roland I think the circle PR gate needs tweaking in github?

**Comment by francois-codes** [comment]
Yes, I will check that

File: circle.yml (hunk 2 above)
Comment (ranmeirman) [reply circle.yml]: https://user-images.githubusercontent.com/5968910/37973854-20e465ae-31e4-11e8-9db1-21e167dd717b.png

Comment (ranmeirman) [reply circle.yml]: checkout circle ci docs and specific tip:
https://circleci.com/docs/2.0/caching/#clearing-cache

**Comment by dickie81** [comment]
@ranmeirman, @f-roland what is outstanding that is required to get this merged?

**Review by francois-codes — APPROVED** [review]
LGTM

**Review by ranmeirman — APPROVED** [review]

**Pull request merged**